### PR TITLE
Run integration tests separately from unit tests

### LINKED ISSUE #18112: Separate Integration tests into two separate make commands.
Currently running `make test` runs a set of integration tests (covering RPC endpoints in Status Go) and a set of unit tests covering reframe subs and events etc.

We should split these into two separate jobs. This will make it clearer and also allow these tests to be run externally, there is currently some discussion about running these as part of Status Go pr's and this will be needed to make that happen.

This might involve extending the current CI job in our codebase to handle this change.

### PR DESCRIPTION
**Pull request by ilmotta**
Fixes https://github.com/status-im/status-mobile/issues/18112

## Summary

Run Clojure integration tests separately from unit tests in the CI.

### Why?

1. Our integration tests generate a lot more log and noise than unit tests. When a unit or integration test fails, it's harder for the developer to spot what failed. Having separate steps in the pipeline makes this process easier.
2. During development, some devs run unit and integration tests using `make test-watch`, which re-runs the slow integration tests alongside unit tests. The dev can control the namespace regex that decides which test namespaces to run, but there's no easy way to watch only unit tests or only integration tests.

### Changelog:

- `make test` still exists, so if you have been using it, as well as `make test-watch`, they should all work exactly the same.
- `[Changed]` As part of the check stage, Jenkins will run `Lint` and `Unit Tests` in parallel. `Integration Tests` run later because running them at the same time as `Unit Tests` caused errors (see https://ci.status.im/blue/organizations/jenkins/status-mobile%2Fprs%2Ftests/detail/PR-18304/2/pipeline).
- `[Added]` `make unit-test` and `make unit-test-watch` run unit tests only. Watching all unit tests is faster now because we ignore integration tests and we only compile shadow-cljs `:mock` target once. We are at approximately 10-15s to re-run all unit tests after saving a watched file, depending on your hardware. If you change `mocks.js_dependencies.cljs`, you must re-run the make target.
- `[Added]` `make integration-test` and `make integration-test-watch` run integration tests only. These are much slower than the unit tests.

Original idea from @J-Son89 :+1:

#### Areas that may be impacted

Most definitely nothing for production code.

### Steps to test

Of course the Jenkins output since we're changing the pipeline, but you can also run the new make targets locally to check.

status: ready

### CODE REVIEW

**Review by siddarthkay — APPROVED** [review]
looks good to me!

**Review by ajayesivan — APPROVED** [review]
Nice work 🎉 !

File: Makefile
Comment (yakimant) [inline Makefile]: Code duplication with `unit-test` and `integration-test`.

I would move to a shared task or script.

File: Makefile
Comment (yakimant) [inline Makefile]: Same, code duplication. Better to extract and parametrise.

**Review by yakimant — CHANGES_REQUESTED** [review]

**Comment by yakimant** [comment]
Good job! Wish status-go unit and integration tests are also split one day!

File: Makefile
Comment (ilmotta) [reply Makefile]: @yakimant, I agree, and I was anticipating this kind of comment in this PR. Since you rejected the PR I will refactor. Thanks for the feedback 👍🏼 

[Just my opinion] I personally don't like makefiles to create interfaces for development, it's always the same story, they end up being a layer of indirection to scripts. Now, using make to help efficiently build source code is a different story, but that's not what our targets do. I worked in a few projects that used makefiles only for that purpose and the rest were all scripts following some common conventions. It was pretty nice! Eventually one project evolved to have an internal CLI, but okay, that can be overkill for us now.

**Review by vkjr — APPROVED** [review]

File: Makefile
Comment (ilmotta) [reply Makefile]: @yakimant, please, have a look at commit to give your second review pass https://github.com/status-im/status-mobile/pull/18304/commits/9c3459aaf1559e445a9b0f32ab0b81b61c0010ac. I could remove a substantial amount of code duplication between all clojure test targets. The small duplication between default target variables seems alright to me, but is that good enough for you?

File: Makefile
Comment (yakimant) [inline Makefile]: What if you move this to subtasks too?
Will it work?

File: Makefile
Comment (yakimant) [reply Makefile]: Thanks @ilmotta, I would move `export TARGET := clojure` it possible too.
If not - I'm ok to merge.

File: Makefile
Comment (yakimant) [reply Makefile]: Why don't you like make for the purpose of interface for usefull scripts?
I find it useful - you just type `make` and you see all the options.

What would be the alternative?

You can always place your code to the scripts and call them directly if you don't want to use make.

File: Makefile
Comment (ilmotta) [reply Makefile]: Good idea and it does work locally. Pushed now, hopefully it works in the CI too.

Edit: All green ;)

**Review by yakimant — APPROVED** [review]
Looks much better now, thanks!

File: Makefile
Comment (ilmotta) [reply Makefile]: Thanks for asking @yakimant. First of all, I don't feel strong about replacing `make`, but I felt like sharing anyway. Subconsciously I probably just wanted to hear your perspective, which I do in fact :)

> What would be the alternative?

The alternative could be `just` (https://github.com/casey/just), which is described as a *command runner* (which is what we use `make` for in status-mobile). I've been trying to use it in personal projects to get the feeling of the good and bad things. So far, it's quite smooth coming from make.

> I find it useful - you just type make and you see all the options.

Do you mean completion at point? Yes, that's very useful. Interestingly, our Makefile contains cryptic code to support `make help`, that comes OTB with `just`.

Another alternative for a group of Clojure devs is to replace scripts and/or some/all make targets with babashka behind a CLI (for discoverability). This wouldn't fly well for others that don't know Clojure. It could be in another language. As I said before, it sounds overkill, but it could promote a better DX over time (which was the case in my past experience)

In my systems I don't even have `make` available by default, although it's an essential piece of software, nowadays I also install `just`.

> You can always place your code to the scripts and call them directly if you don't want to use make

Yes, but then it's like I said, from my personal perspective, `make` becomes an unnecessary piece of software when it's used mostly as an indirection layer to scripts.

If we only used scripts and a basic CLI for discoverability (optional) it would work just as well, or at lest that's my theory :)

**Review by J-Son89 — APPROVED** [review]
🙌

**Comment by J-Son89** [comment]
@cammellos - is it possible to make this part of the CI process for Status Go afterwards? I don't mean technically as I'm sure we can figure that out, but more that we have agreement to add it? 🤔

**Comment by yakimant** [comment]
@J-Son89, that's interesting! Do you want to run `status-mobile` tests on each `status-go` changes? Would be cool to run `status-desktop` tests too.

**Comment by J-Son89** [comment]
Yep exactly. Essentially we have contract tests then for the mobile client. Would be great if Desktop did the same too! 🙏

File: ci/Jenkinsfile.tests
Comment (jakubgs) [inline ci/Jenkinsfile.tests:65]: Can those not run in parallel?

File: Makefile
Comment (jakubgs) [inline Makefile:342]: Why does this start with underscore? And it's missing a help message.

File: Makefile
Comment (jakubgs) [inline Makefile]: Why does this start with underscore? And it's missing a help message.

File: Makefile
Comment (jakubgs) [reply Makefile:342]: And the name should be `test-clojure` if we want to have consistent prefixes.

File: Makefile
Comment (jakubgs) [reply Makefile]: And should be called `test-clojure-watch` or similar to have the same prefix as other test targets.

File: Makefile
Comment (jakubgs) [inline Makefile]: Shouldn't this be called `test-all` then for clarity?

File: Makefile
Comment (jakubgs) [reply Makefile]: Also, why do we need a separate target? Why can't you just have a variable that decides if it's a watch version or not?

**Review by jakubgs — CHANGES_REQUESTED** [review]
I think we are creating too many targets. If you want `-watch` targets you can just make the normal target accept a `WATCH`(or some other name) parameter and then just create aliases using it:
```Makefile
test: WATCH ?= false
ifeq ($(WATCH), false)
	some thing
else
	some other thing
endif

test-watch: WATCH := true
test-watch: test

File: Makefile
Comment (ilmotta) [reply Makefile:342]: > Why does this start with underscore? And it's missing a help message.

There are other targets prefixed with underscore, so I used the same idea to mean "private" targets that shouldn't be used directly. Edit: Being "private", I preferred not to document them, but I can add something no problem.

> And the name should be test-clojure if we want to have consistent prefixes.

Good point, will do.

File: Makefile
Comment (ilmotta) [reply Makefile]: I wanted to keep the code more backwards compatible, so other devs not aware of these changes can keep their workflows.

File: Makefile
Comment (ilmotta) [reply Makefile]: I think it's a matter of convenience @jakubgs. Passing parameters to make is not as convenient as using the auto-complete of available targets and just pressing `tab` and `enter`. Obviously, this leads to a bit of code duplication.

We can have a variable for sure, a matter of taste :)

File: ci/Jenkinsfile.tests
Comment (ilmotta) [reply ci/Jenkinsfile.tests:65]: I mentioned this in the PR description:

> Integration Tests run later because running them at the same time as Unit Tests caused errors (see https://ci.status.im/blue/organizations/jenkins/status-mobile%2Fprs%2Ftests/detail/PR-18304/2/pipeline)

Originally, I wanted to run them all in parallel. I ran the pipeline twice and faced that error and honestly didn't care about solving the problem, not now at least. It's a good question nonetheless.

The same question can be made about the component tests, they are currently not running in parallel with the other checks and I don't know why.

**Comment by ilmotta** [comment]
> I think we are creating too many targets. If you want `-watch` targets you can just make the normal target accept a `WATCH`(or some other name) parameter and then just create aliases using it:
> 
> ```makefile
> test: WATCH ?= false
> ifeq ($(WATCH), false)
> 	some thing
> else
> 	some other thing
> endif
> 
> test-watch: WATCH := true
> test-watch: test
> ```

@jakubgs, unfortunately the code you suggested doesn't work. The conditional `ifeq` only takes effect if we call make from make (as in `$(MAKE)`), not when `test` is a prerequisite of `test-watch`.

Here's an example:

```makefile
foo: export TARGET := clojure
foo: export WATCH ?= false
foo:
ifeq ($(WATCH), true)
	@echo watch=true
else
	@echo watch=false
endif

bar: export WATCH := true
bar: foo
```

```sh
make foo            # => prints watch=false
make bar            # => prints watch=false
make foo WATCH=true # => prints watch=true
make bar WATCH=true # => prints watch=true
```

If we change to recursive make call it works.

```makefile
bar:
	$(MAKE) foo WATCH=true
```

So as we can see the variable doesn't propagate. Therefore, the solution is to either drop these targets `*-watch` and let the dev pass `WATCH=true` or call make from make. I'm leaning on the side of removing `test-unit-watch` and `test-integration-watch` and rely on the dev to pass `WATCH=true`.

File: Makefile
Comment (ilmotta) [reply Makefile]: Resolved by https://github.com/status-im/status-mobile/pull/18304/commits/bcf0af0a8544af4564e510affb8a50969015cf04

Now the dev must pass `WATCH=true` to the test targets they care about watching files and the Makefile is more DRY now.

Also renamed the targets to be more consistent. I did not rename the component test targets. I consider that out of scope for this PR.

File: Makefile
Comment (ilmotta) [reply Makefile:342]: Resolved by https://github.com/status-im/status-mobile/pull/18304/commits/bcf0af0a8544af4564e510affb8a50969015cf04

I kept the underscore prefix like other targets and I didn't add any documentation. The "public" targets have a description though. Is that okay for you?

File: ci/Jenkinsfile.tests
Comment (ilmotta) [reply ci/Jenkinsfile.tests:65]: @jakubgs, I would like to keep the integration tests separate from unit test because I don't want to dedicate even more time to this PR. In the future somebody can figure out if & how we can run integration and component tests in parallel with unit tests and linting.

Comment (jakubgs) [reply ci/Jenkinsfile.tests:65]: That makes sense.

File: Makefile
Comment (jakubgs) [reply Makefile]: Fair enough.

**Review by jakubgs — APPROVED** [review]
@ilmotta thanks for testing that. That is indeed annoying, `make` is such an obtuse tool to use.

I guess adding a `test-watch` target that uses the `$(MAKE)` trick is fine for the sake of discoverability of such command in the `make` help message.

**Pull request merged**
